flow/enhancers/socket_info.go

```diff
@@ -0,0 +1,371 @@
+			si.Pid = pid
+			si.Tid = tid
+		}
+	}
+
+	// Retreive ppid and task name
+	stat, err := ioutil.ReadFile(procProcessRoot + "/stat")
+	if err != nil {
+		return
+	}
+	var pid, ppid, status int
+	var name string
+	fmt.Sscanf(string(stat), "%d %s %c %d", &pid, &name, &status, &ppid)
+	if si.Pid == 0 {
+		si.Pid = pid
+	}
+	si.Name = name[1 : len(name)-1]
+	si.PPid = ppid
+}
+
+func getPIDs(path string) []string {
+	pids := []string{}
+	proc, err := os.Open(path)
+	if err != nil {
+		return pids
+	}
+	defer proc.Close()
+	names, err := proc.Readdirnames(0)
+	if err != nil {
+		return pids
+	}
+	for _, n := range names {
+		_, err := strconv.ParseInt(n, 0, 0)
+		if err != nil {
+			continue
+		}
+		pids = append(pids, n)
+	}
+	return pids
+}
+
+func getFDs(path string) []string {
+	fds := []string{}
+	proc, err := os.Open(path + "/fd")
+	if err != nil {
+		return fds
+	}
+	defer proc.Close()
+	names, err := proc.Readdirnames(0)
+	if err != nil {
+		return fds
+	}
+	for _, n := range names {
+		fd, err := os.Readlink(path + "/fd/" + n)
+		if err != nil {
+			continue
+		}
+		fds = append(fds, fd)
+	}
+	return fds
+}
+
+var rFdSocket = regexp.MustCompile("socket:\\[(\\d+)\\]$")
+
+func (s *SocketInfoEnhancer) scanProc() {
+	s.resetScanProcNet()
+
+	pids := getPIDs("/proc")
+	for _, p := range pids {
+		procTask := "/proc/" + p + "/task"
+		tasks := getPIDs(procTask)
+		for _, t := range tasks {
+			process := procTask + "/" + t
+			s.updateProcNet(process)
+
+			fds := getFDs(process)
+			for _, f := range fds {
+				inode := rFdSocket.FindStringSubmatch(f)
+				if len(inode) == 0 {
+					continue
+				}
+
+				var isUDP = true
+				sinfo, ok := s.cacheUDP.Get(inode[1])
+				if !ok {
+					isUDP = false
+					sinfo, ok = s.cacheTCP.Get(inode[1])
+					if !ok {
+						continue
+					}
+				}
+				si := sinfo.(SocketInfo)
+				parseProcStat(process, &si)
+				if isUDP == true {
+					s.cacheUDP.Set(inode[1], si, cache.DefaultExpiration)
+				} else {
+					s.cacheTCP.Set(inode[1], si, cache.DefaultExpiration)
+				}
+			}
+		}
+	}
+}
+
+func ipToHex(ip46 net.IP) string {
+	s := ""
+	ip := ip46.To4()
+	if ip == nil {
+		ip = ip46
+	}
+	for i := range ip {
+		s += fmt.Sprintf("%02X", ip[(len(ip)-1)-i])
+	}
+	return s
+}
+
+func (s *SocketInfoEnhancer) getSocketInfoLocal(f *flow.Flow) bool {
+	ip := net.ParseIP(f.Network.A)
+	if ip == nil {
+		return false
+	}
+
+	port, _ := strconv.ParseUint(f.Transport.A, 10, 0)
+	ipPortA := fmt.Sprintf("%s:%04X", ipToHex(ip), port)
+	ipPortB := ""
+
+	ip = net.ParseIP(f.Network.B)
+	if ip != nil {
+		port, _ = strconv.ParseUint(f.Transport.B, 10, 0)
+		ipPortB = fmt.Sprintf("%s:%04X", ipToHex(ip), port)
+	}
+
+	retried := false
+retry:
+	var cache *cache.Cache
+	if f.Transport.Protocol == flow.FlowProtocol_UDPPORT {
+		cache = s.cacheUDP
+	} else if f.Transport.Protocol == flow.FlowProtocol_TCPPORT {
+		cache = s.cacheTCP
+	}
+
+	var siA, siB bool
+	items := cache.Items()
+	for _, item := range items {
```

Comment:
iterating over all the cache items seems to be a bit overkill. I would prefer to better index what we looking for. This will maybe allow to keep only one cache instance. We could also add the result of SocketInfoToFlowSocketInfo in the case instead of si